### PR TITLE
#167735038 email password reset

### PR DESCRIPTION
**Pull request by prieste**
**What does this PR do?**

 Resets Users password using their email

  **Description of Task to be completed?**

 - Send an email to the provided email address, from the 'forget_password' post route
 - Accept a new password and token from the body and url path of the 'reset_password' patch route
-  If the token is valid update the password on the db

  **How should this be manually tested?**

 -  startup the server
- on postman enter 'localhost:3000/api/v1/auth/forgot_password' on a post route with a (valid?) email address
- copy the token received and append it to the following url  'localhost:3000/api/v1/auth/reset_password/{token}

  **Any background context you want to provide?**

 N/A

  **What are the relevant stories?**
  pivotal tracker - #167735038

  **Screenshots (if appropriate)**
 N/A
  **Questions:**

 N/A

### CODE REVIEW

**Comment by prieste** [comment]
Current failing checks are for uncovered lines

File: src/controllers/authController.js [hunk 1]
```diff
@@ -16,4 +21,85 @@ export default class AuthController {
...
+  }
+
+  /**
+   * @param {object} req request body
+   * @param {object} res response body
+   * @returns  {object} message
+   */
+  static async forgotPassword(req, res) {
+    const { body: { email } } = req;
+    const result = await User.findOne({ where: { email } });
```
Comment (SamuelKoroh) [inline src/controllers/authController.js]: Remove querying of the user the from controller take it to AuthService

File: src/controllers/authController.js [hunk 2]
```diff
@@ -16,4 +17,32 @@ export default class AuthController {
...
+
+  /**
+   * @param {object} req request body
+   * @param {object} res response body
+   * @returns  {object} message
+   */
+  static async forgotPassword({ body: { email } }, res) {
+    try {
+      const data = await AuthService.forgotPassword(email);
+      res.status(200).json({ status: 'success', data });
```
Comment (Annmary12) [inline src/controllers/authController.js]: Leave a line after declaring variables

File: src/controllers/authController.js [hunk 3]
```diff
@@ -16,4 +17,32 @@ export default class AuthController {
...
+
+  /**
+   * @param {object} req request body
+   * @param {object} res response body
+   * @returns  {object} message
+   */
+  static async resetPassword({ params: { token }, body: { password } }, res) {
+    try {
+      const message = await AuthService.resetPassword(token, password);
+      res.status(200).json({ status: 'success', message });
```
Comment (Annmary12) [inline src/controllers/authController.js]: Leave a line after declaring variables

File: src/database/config.js [hunk 4]
```diff
@@ -1,13 +1,14 @@
 require('dotenv').config();
 
+const username = process.env.DBUSERNAME || process.env.USERNAME;
 module.exports = {
   development: {
-    username: process.env.USERNAME,
+    username,
```
Comment (Annmary12) [inline src/database/config.js]: why did you change line 6?

Comment (Annmary12) [inline src/database/config.js]: why did you change line 6?

File: src/services/authService.js [hunk 5]
```diff
@@ -29,4 +31,47 @@ export default class AuthService {
...
+  static async forgotPassword(email) {
+    const result = await User.findOne({ where: { email } });
+    if (!result) throw new Error('Email not found');
+    const { dataValues: { first_name: name } } = result;
+    const token = await jwt.sign({
+      email
+    }, jwtSecret, { expiresIn: '1h' });
+    const emailDetails = {
+      receiver: email,
+      sender: 'chubi.best@gmail.com',
```
Comment (Annmary12) [inline src/services/authService.js]: The sender details should be stored in the `.env` file because it can be changed.

File: src/services/authService.js [hunk 6]
```diff
@@ -29,4 +31,47 @@ export default class AuthService {
...
   }
+
+  /**
+ *
+ * @param {string} email email address
+ * @returns {string} token
+ */
+  static async forgotPassword(email) {
+    const result = await User.findOne({ where: { email } });
+    if (!result) throw new Error('Email not found');
```
Comment (Annmary12) [inline src/services/authService.js:43]: Please always remember to leave a line after declaring a  variable or group of variables. There should be a line in between `41` and `42`, `42` and `43`,  `53` and `54` .
Let this style reflect in all your codebase Please...

File: src/services/authService.js [hunk 7]
```diff
@@ -29,4 +31,47 @@ export default class AuthService {
...
+  }
+
+  /**
+   *
+   * @param {string} token request token
+   * @param {string} password new password
+   * @returns {string} message
+   * */
+  static async resetPassword(token, password) {
+    const { email } = jwt.verify(token, jwtSecret);
```
Comment (Annmary12) [inline src/services/authService.js]: There is no check to ensure that the token belongs to the user or if the token has expired.

File: src/services/mailer/index.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import dotenv from 'dotenv';
+import sgMail from '@sendgrid/mail';
+
+dotenv.config();
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+
+const templates = {
+  verify_email: 'd-1903b0acf59940cea2618e012f891007',
+  reset_password: 'd-9fdf0f733e28408a8dfc06fbebf74070'
```
Comment (Annmary12) [inline src/services/mailer/index.js]: Line 8 and 9 values should be set in our .env file for security purposes

File: src/services/mailer/index.js [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    templateId: templates[data.templateName],
+
+    dynamic_template_data: {
+      name: data.name,
+      verify_email: data.confirm_account__url,
+    }
+  };
+  try {
+    await sgMail.send(message);
+  } catch (e) {
```
Comment (Annmary12) [inline src/services/mailer/index.js]: Please change `e` to `error`

File: src/test/passwordRest.spec.js [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+
+describe('/api/v1/auth', () => {
+  let passwordResetToken;
+
+  before((done) => {
+    TestHelper.destroyModel('User');
+    done();
+  });
+  describe('POST /forgot_password', () => {
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js:26]: leave a line after line 18

File: src/test/passwordRest.spec.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    before(async () => {
+      await db.User.create({
+        first_name: 'irellevant',
+        last_name: 'Tester',
+        email: 'testa@test.com',
+        password: 'PasswordTest123'
+      });
+    });
+    it('should return an error for an unregistered user', async () => {
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js:27]: leave a line after line 27

File: src/test/passwordRest.spec.js [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+      const { status, text } = await chai.request(app)
+        .post('/api/v1/auth/forgot_password')
+        .send({
+          email: 'yahan@mail.com'
+        });
+      expect(status).to.equal(400);
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js:34]: leave a line after line 33

File: src/test/passwordRest.spec.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+      expect(JSON.parse(text).error).to.equal('Email not found');
+    });
+    it('should send a message, token and an email for a registered user', async () => {
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js:38]: leave a line after line 35

File: src/test/passwordRest.spec.js [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+      const { status, text } = await chai.request(app)
+        .post('/api/v1/auth/forgot_password')
+        .send({
+          email: 'testa@test.com'
+        });
+      passwordResetToken = JSON.parse(text).data;
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js]: leave a line after line 41 and 42

File: src/test/passwordRest.spec.js [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+      expect(JSON.parse(text).data).to.equal(passwordResetToken);
+      expect(status).to.equal(200);
+    });
+  });
+  describe('POST /resetPassword/:token', () => {
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js]: leave a line after line 46

File: src/test/passwordRest.spec.js [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+    it('should return a message on succesful update', async () => {
+      const { status, text } = await chai.request(app)
+        .patch(`/api/v1/auth/reset_password/${passwordResetToken}`)
+        .send({
+          password: 'password'
+        });
+      expect(status).to.equal(200);
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js]: leave a line after line 54

File: src/test/passwordRest.spec.js [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+      expect(JSON.parse(text).message).to.equal('Password changed');
+    });
+    it('should return an error for an invalid token', async () => {
```
Comment (Annmary12) [inline src/test/passwordRest.spec.js:63]: Please leave a line after every `it` block and after a `variable` is declared for readability sake.

File: src/database/config.js (hunk 4 above)
Comment (prieste) [reply src/database/config.js]: was a workaround for some issues I had with my machine,  I'll undo it

File: src/test/passwordRest.spec.js (hunk 10 above)
Comment (prieste) [reply src/test/passwordRest.spec.js:26]: Eslint flags it when I try to leave a line, should I disable it?

File: src/test/passwordRest.spec.js (hunk 11 above)
Comment (prieste) [reply src/test/passwordRest.spec.js:27]: Eslint flags it when I try to leave a line, should I disable it?

File: src/test/passwordRest.spec.js (hunk 13 above)
Comment (prieste) [reply src/test/passwordRest.spec.js:38]: Eslint flags it when I try to leave a line, should I disable it?

File: src/test/passwordRest.spec.js (hunk 11 above)
Comment (Annmary12) [reply src/test/passwordRest.spec.js:27]: I am not sure of that because I gave this feedback to other members of the team and they implemented without Eslint complaining

Comment (Annmary12) [reply src/test/passwordRest.spec.js:27]: I just pulled your branch to my local system and tried it but Eslint didn't complain

File: src/services/authService.js (hunk 7 above)
Comment (prieste) [reply src/services/authService.js]: The verify method throws an error for an invalid token that is handled in the controller, the email was signed upon hitting the first route, so the verify method returns that email provided by the user in the first place... perhaps I need more context to understand what you want me to do

**Comment by Annmary12** [comment]
@prieste , resolve conflict on your PR. Ensure you and @amoskeyz are using the same email templating format

File: package.json [hunk 18]
```diff
@@ -6,6 +6,7 @@
   "scripts": {
     "db:migration": "cross-env NODE_ENV=test sequelize db:migrate",
     "undo:migration": "cross-env NODE_ENV=test sequelize db:migrate:undo:all",
+    "test-dev": "NODE_ENV=test sequelize db:migrate:undo:all && NODE_ENV=test sequelize db:migrate && npm test",
```
Comment (Annmary12) [inline package.json]: Why did you remove our `test` script? please check the development branch to see what we have.

File: package.json [hunk 19]
```diff
@@ -25,11 +26,13 @@
...
     "@babel/register": "^7.5.5",
     "@babel/runtime": "^7.5.5",
-    "bcrypt": "^3.0.6",
     "@hapi/joi": "^15.1.1",
+    "@sendgrid/mail": "^6.4.0",
+    "bcrypt": "^3.0.6",
     "body-parser": "^1.18.3",
     "chai": "^4.2.0",
     "chai-http": "^4.3.0",
+    "chalk": "^2.4.2",
```
Comment (Annmary12) [inline package.json]: Why did you install `chalk` and what are you using it for?

File: src/services/authService.js [hunk 20]
```diff
@@ -30,11 +32,73 @@ export default class AuthService {
...
+    const { dataValues: { first_name: name, } } = result;
+    const token = await jwt.sign({
+      email
+    }, jwtSecret, { expiresIn: '1h' });
+    const emailDetails = {
+      receiver: email,
+      sender: process.env.SENDER,
+      templateName: 'reset_password',
+      name,
+      url: `provide_newpass/reset_password/${token}`
```
Comment (Annmary12) [inline src/services/authService.js]: You don't need to add `provide_newpass` in the url, this is fine `reset_password/${token}`

File: src/services/authService.js [hunk 21]
```diff
@@ -30,11 +32,73 @@ export default class AuthService {
...
+   * @param {string} token request token
+   * @param {string} password new password
+   * @returns {string} message
+   * */
+  static async resetPassword(token, password) {
+    const { email } = await jwt.verify(token, jwtSecret);
+    const { dataValues: { reset_token: tokenSaved } } = await User.findOne({ where: { email } });
+
+    if (token !== tokenSaved) throw new Error('Invalid token');
+    if (password === undefined || password.trim().length < 8) throw new Error('Provide valid password');
```
Comment (Annmary12) [inline src/services/authService.js]: You should use our validation schema to check password exist and the length of the password. @cleave3 has implemented this on his task, reach out to him to guide you

File: src/services/authService.js [hunk 22]
```diff
@@ -30,11 +32,73 @@ export default class AuthService {
...
+
+    const newPassword = await Helper.encryptor(password);
+
+    await User.update({
```
Comment (Annmary12) [inline src/services/authService.js]: I noticed that you have this `User.update` function in two places. Can you abstract this to an authService class. we can have something that looks like this in the authService class.

```
static async updateUser(query,options) {
   User.update({options}, {
    where: {
       query
     })
}
```

You can use this in your controller or any where.
Example with your code below:

```
const options = {
      password: newPassword,
      reset_token: null
}

const query = email;

// then call the function

await authService.updateUser(query, options)
```

File: src/services/emailService.js [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+dotenv.config();
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+
+const templates = {
+  verify_email: process.env.VERIFY_TEMPLATE,
+  reset_password: process.env.RESET_TEMPLATE
+};
+
+export const sendMail = async (data) => {
```
Comment (Annmary12) [inline src/services/emailService.js:18]: Please use Jsdocs here

File: src/services/authService.js [hunk 24]
```diff
@@ -30,11 +32,65 @@ export default class AuthService {
...
+    const { dataValues: { first_name: name, } } = result;
+    const token = await jwt.sign({
+      email
+    }, jwtSecret, { expiresIn: '1h' });
+    const emailDetails = {
+      receiver: email,
+      sender: process.env.SENDER,
+      templateName: 'reset_password',
+      name,
+      url: `reset_password/${token}`
```
Comment (Annmary12) [inline src/services/authService.js]: This url should be `${process.env.APP_URL}/reset_password/${token}`
The `process.env.APP_URL` can be anything for now but will change when we start working on our frontend. @amoskeyz did something like this

**Comment by Annmary12** [comment]
@amoskeyz , I couldn’t test your feature on swagger, the signup is breaking and if you check our signup with the staging link, you will notice that is working properly. I believe is as the result of your changes on account verification task. Please fix this and test it with swagger to ensure everything is working properly.

File: src/controllers/authController.js (hunk 1 above)
Comment (prieste) [reply src/controllers/authController.js]: Okay will do
